win: port test obj_sync

JLLLinn opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/1277)
<!-- Reviewable:end -->

wlemkows commented:
Review status: 0 of 30 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/windows/win_file.c, line 142 at r2](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTrZziLb6VBuIwEiQ1n:-KTrZziLb6VBuIwEiQ1o:b1gs8ak) ([raw file](https://github.com/pmem/nvml/blob/c0060afdfd4049b0e01ab65a7d48c64bef2c8b54/src/windows/win_file.c#L142)):*
> ```C
> 
> void
> clock_gettime(clock_t id, struct timespec *ts)
> ```

I added implementation for CLOCK_MONOTONIC and moved this function to win_time file.
https://github.com/pmem/nvml/pull/1282/files
Please move your impl too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

wlemkows commented:
Reviewed 1 of 30 files at r1.
Review status: 1 of 30 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

JLLLinn commented:
Review status: 1 of 32 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/windows/win_file.c, line 142 at r2](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTrZziLb6VBuIwEiQ1n:-KTu9-xA_5UbCvsDBvwA:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/c0060afdfd4049b0e01ab65a7d48c64bef2c8b54/src/windows/win_file.c#L142)):*
<details><summary><i>Previously, wlemkows (Weronika Lemkowska) wrote…</i></summary>
> I added implementation for CLOCK_MONOTONIC and moved this function to win_time file.
> https://github.com/pmem/nvml/pull/1282/files
> Please move your impl too.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 26 of 30 files at r1, 2 of 4 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test/obj_sync/obj_sync.c, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuJHtTu9-C1xM_7Bhy:-KTuJHtTu9-C1xM_7Bhz:b-44cp9r) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.c#L101)):*
> ```C
> {
> #ifdef _WIN32
> 	__sync_fetch_and_add64(&pop->run_id, 2);
> ```

What about the same approach as for __sync_bool_compare_and_swap()?
Define util_fetch_and_addXX() wrappers in util.h, and use util_fetch_and_add64() here. (no #ifdef)

---

*[src/test/obj_sync/obj_sync.vcxproj, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuK51MScHZe2jRawfj:-KTuK51MScHZe2jRawfk:bae5l9i) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L65)):*
> ```XML
>       <ForcedIncludeFiles>platform.h;mocks_windows.h</ForcedIncludeFiles>
>       <CompileAs>CompileAsC</CompileAs>
>       <DisableSpecificWarnings>4028</DisableSpecificWarnings>
> ```

Is this really needed?

---

*[src/test/obj_sync/obj_sync.vcxproj, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuOjXMgjaNBNgWpRd2:-KTuOjXMgjaNBNgWpRd3:bz7mx8c) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L101)):*
> ```XML
>   </ItemDefinitionGroup>
>   <ItemGroup>
>     <ClCompile Include="..\..\common\file.c" />
> ```

Are all these files really required by this test? Isn't sync.c enough?

---

*[src/windows/include/platform.h, line 183 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuSFQE9DYbEsxAT-ZN:-KTuSFQE9DYbEsxAT-ZO:b-mi1sgb) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/windows/include/platform.h#L183)):*
> ```objc
> #ifndef DELTA_WIN2UNIX
> /* number of useconds between 1970-01-01T00:00:00Z and 1601-01-01T00:00:00Z */
> #define DELTA_WIN2UNIX (11644473600000000ull)
> ```

We have it already in pthread_windows.c. Perhaps, we could get rid of one of those?
Also, if it's used in win_time.c only, then why not to define it there?

---

*[src/windows/liblinux/win_time.c, line 38 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuQlp46irVqncbaYH_:-KTuQlp46irVqncbaYHa:b-6oleat) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/windows/liblinux/win_time.c#L38)):*
> ```C
>  */
> 
> int
> ```

missing brief

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

JLLLinn commented:
Review status: 28 of 32 files reviewed at latest revision, 6 unresolved discussions.

---

*[src/test/obj_sync/obj_sync.c, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuJHtTu9-C1xM_7Bhy:-KTvDbQrLZB9Q-ZjqUt_:bbcv1vh) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.c#L101)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What about the same approach as for __sync_bool_compare_and_swap()?
> Define util_fetch_and_addXX() wrappers in util.h, and use util_fetch_and_add64() here. (no #ifdef)
</details>
To do this we will likely have to change the feature code (all __sync_bool_compare_and_swap to util_bool_compare_and_swap) as happens for util_fetch_and_add, should I proceed with it?

---

*[src/test/obj_sync/obj_sync.vcxproj, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuK51MScHZe2jRawfj:-KTvD7rb68xiR75yEXmQ:bbnolh3) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Is this really needed?
</details>
VS warns about this since pthread_mutex_init (and 2 others) has "__restrict" in their definition but not in the FUNC_MOCK

---

*[src/test/obj_sync/obj_sync.vcxproj, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuOjXMgjaNBNgWpRd2:-KTvCJ197SQKsQDzI_uZ:bkwwldr) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L101)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Are all these files really required by this test? Isn't sync.c enough?
</details>
Done. set.c is needed since a mocked function is called during pre-loading. also need pthread_windows for all the real function definition.  Others are excluded after libpmemobj is added as reference.

---

*[src/windows/include/platform.h, line 183 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuSFQE9DYbEsxAT-ZN:-KTvCCtdxstHnJCrsjgN:bjotsuz) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/windows/include/platform.h#L183)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> We have it already in pthread_windows.c. Perhaps, we could get rid of one of those?
> Also, if it's used in win_time.c only, then why not to define it there?
</details>
Moved this to win_time, Done

---

*[src/windows/liblinux/win_time.c, line 38 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuQlp46irVqncbaYH_:-KTvCB4ot1ia8NgEXmkf:b-896fix) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/windows/liblinux/win_time.c#L38)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> missing brief
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

lplewa commented:
Review status: 27 of 32 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/obj_sync/obj_sync.vcxproj, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuK51MScHZe2jRawfj:-KTxQUG95p8eAtOU7LM_:b-da4hi3) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L65)):*
<details><summary><i>Previously, JLLLinn (Jiaxin) wrote…</i></summary>
> VS warns about this since pthread_mutex_init (and 2 others) has "__restrict" in their definition but not in the FUNC_MOCK
</details>
but can we change this to #pragma warning in yours windows only file(with appropriate comment)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 1 of 30 files at r1, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/obj_sync/obj_sync.c, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuJHtTu9-C1xM_7Bhy:-KTyEa8Il28Skh0Yt_rm:b-jcu0pb) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.c#L101)):*
<details><summary><i>Previously, JLLLinn (Jiaxin) wrote…</i></summary>
> To do this we will likely have to change the feature code (all __sync_bool_compare_and_swap to util_bool_compare_and_swap) as happens for util_fetch_and_add, should I proceed with it?
</details>
Right. It's not a big deal to make such a change. Could be a separate patch, if you prefer.

---

*[src/test/obj_sync/obj_sync.vcxproj, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuK51MScHZe2jRawfj:-KTyDsamxw0igIsBXPYu:b-ini83r) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L65)):*
<details><summary><i>Previously, lplewa (Łukasz Plewa) wrote…</i></summary>
> but can we change this to #pragma warning in yours windows only file(with appropriate comment)
</details>
Nope. IMHO, the proper solution is to add "__restrict" to the FUNC_MOCK, like this:
```c
FUNC_MOCK(pthread_mutex_init, int,
                pthread_mutex_t *__restrict mutex,
                const pthread_mutexattr_t *__restrict attr)
...
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

JLLLinn commented:
Review status: 28 of 32 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/test/obj_sync/obj_sync.c, line 101 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuJHtTu9-C1xM_7Bhy:-KU-M79TCY8XbLPbGEZI:bw0mt3k) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.c#L101)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Right. It's not a big deal to make such a change. Could be a separate patch, if you prefer.
</details>
Thank you, Once this is merged I will make a separate patch.

---

*[src/test/obj_sync/obj_sync.vcxproj, line 65 at r3](https://reviewable.io:443/reviews/pmem/nvml/1277#-KTuK51MScHZe2jRawfj:-KU-MEpEXpKXKbXixOLZ:b-4uj44h) ([raw file](https://github.com/pmem/nvml/blob/dbf7aee0a4a8060a434c35a28ee7db7e24d61e83/src/test/obj_sync/obj_sync.vcxproj#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Nope. IMHO, the proper solution is to add "__restrict" to the FUNC_MOCK, like this:
> ```c
> FUNC_MOCK(pthread_mutex_init, int,
>                 pthread_mutex_t *__restrict mutex,
>                 const pthread_mutexattr_t *__restrict attr)
> ...
> ```
> 
> 
</details>
Done. It is still processing but it passed on my own travis

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277)*
<!-- Sent from Reviewable.io -->

krzycz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/1277#-:-KU0mmDPuYVd5k2Uh_gb:bnfp4nl)*
<!-- Sent from Reviewable.io -->
